### ISSUE TITLE
I've encountered a similar issue. Wondering whether introducing an extra dummy argument to a function can be used to comply with Jax check for the matching shape of the bwd()' output and fwd()'s input, thus allowing it to carry non-gradient information in the backward pass? For example as in the following code:

### ISSUE THREAD

**Issue by BohdanK-1QBit** [google/jax#4324]
I've encountered a similar issue. Wondering whether introducing an extra dummy argument to a function can be used to comply with Jax check for the matching shape of the bwd()' output and fwd()'s input, thus allowing it to carry non-gradient information in the backward pass? For example as in the following code: 

```
from jax import custom_vjp, grad
import jax.numpy as jnp

from jax import custom_vjp, grad
import jax.numpy as jnp

#=======================================
@custom_vjp
def invertible_mult(x, dummy_x):
  return 2*x, dummy_x

def invertible_mult_fwd(x, dummy_x):
  return invertible_mult(x, dummy_x), ()

def invertible_mult_bwd(res, g):
  true_g, x = g 
  true_g = 2.*true_g
  return (true_g, x/2)

invertible_mult.defvjp(invertible_mult_fwd, invertible_mult_bwd)

#=======================================
# used as a bridge between a sequence of invertible calculation followed by
# a regular set of calculations. In the forward pass it acts as an identity
# on the first argument while storing its value to initialize it for the
# propagation in the reverse direction 
@custom_vjp
def end_invertible(x, dummy_x):
  return x

def end_invertible_fwd(x, dummy_x):
  return end_invertible(x, dummy_x), (x)

def end_invertible_bwd(res, g):
  x = res
  return (g, x)

end_invertible.defvjp(end_invertible_fwd, end_invertible_bwd)

#=======================================
def full_transform(x):
    dummy_x = jnp.ones_like(x)
    y, dummy_y = invertible_mult(x, dummy_x)
    y, dummy_y = invertible_mult(y, dummy_y)
    y = end_invertible(y, dummy_y)
    return jnp.sum(y)

#=======================================
grad(full_transform)(jnp.array([1., 5]))
```

In particular, I am not quite sure whether `invertible_mult_bwd() `would still cache some variables behind scenes for potential composite call to grad() transformations? (the ultimate goal here is to reduce memory footprint)

_Originally posted by @BohdanK-1QBit in https://github.com/google/jax/issues/2887#issuecomment-694326322_
